### ISSUE TITLE
Scripts don't work out of the box

### ISSUE THREAD

**Issue by sebastian-c** [UNFAOstatistics/pocketbook_database#1]
Hi Markus, I thought I'd clone this project for educational purposes and came across the following issue. I started with the `pocketbook` project, but realised that this needed to run first.

Following the `run.R` script and after changing the `root.dir`, the following error error is obtained:

```
Error in geogr(x = x, n = 10) : object 'x' not found 
```

This occurs at line 71 and seems to be because `code/FAOSTAT_functions/geogr.R` has the following at line 48:

```
geogr(x = x, n = 10)
```

This, not being defined, fails. Is it meant to be an example (and hence commented out) or is something meant to be sourced beforehand from which the `x` is meant to be inherited? There are a few more functions further down that seem to have a similar issue.

**Comment by muuankarski** [comment]
Hi Sebastian, 

commented out the garbage from yesterdays debuggins session in https://github.com/UNFAOstatistics/pocketbook_database/commit/cdca9566bb2691c2f453f54facd54f3bf0ee4bf5 

These have not been tested in any other environment and there may dependencies to files outside the repo. Not too many, but some for sure. Let me know so I can clarify!

**Comment by sebastian-c** [comment]
I might make a pull request when I have a bit of spare time. Just to be clear, all of the files in `code/FAOSTAT_functions/` are meant to be functions only, right?

**Comment by muuankarski** [comment]
right! basically faostat-package frozen at some point in time with few modification

**Comment by sebastian-c** [comment]
Ok, good. The `~/local_data/faostat` folder - Is there a reason that it points there rather than a directory inside the `pocketbook_database` directory? It seems to me that it would be better if it were created there and then added to the `.gitignore`.

**Comment by muuankarski** [comment]
This is due to my multi-device setup. Only the code & processed data is [rsynced](https://en.wikipedia.org/wiki/Rsync) across devices, not the raw/bulk data that lies in `~/local_data`. Contains ~20-30gb of "idle" data. In a single machine setup that would perhaps be way to go. Of course one can do `if (Sys.info()[["something"]] == "mysetup") a_different_path_for_this_machine` and add it to gitignore if want to keep it in subfolder!

**Comment by sebastian-c** [comment]
It sounds at this point like it might be a good idea to have a config file (maybe as a dcf so as not to have extra deps). Something that looks like:

```
base_dir: "~/projects/pocketbook_database"
local_data_dir: "~/local_data"
```

Then have a function that reads that file at the beginning of the script. Add it to the `.gitignore` so that people can have different ones, but throw an error if it's not there with a suggestion on how to create it. Maybe create a default one if it doesn't exist.

**Comment by muuankarski** [comment]
It would be a good idea and simple to implement. However, this process is a total disaster and fixing it has never been a priority since i took it over for more than two years ago. Work on this has been limited only to "revive" it to spit out the data for producing the pdfs in a hurry. This has not been treated as a piece of software and all good practices of sustainable software development has been ignored here. 

However, for someone interested in sharing the experience, it should be made as easy as possible! I have not been provided any equipment for this work, but this fires up in a fresh installation of R on ubuntu based linux distros "out of the box". Sure you need to have 16+gb of RAM and plenty of time wait for it to run through.

But any kind of attention is certainly welcome. I don't think anyone else has properly looked into the code during the past two years.

**Comment by sebastian-c** [comment]
I understand and I think you've done very well to reanimate it enough so that it can shamble along enough to do what it needs to do. I can't make any kind of official time for this, but in my downtime I'd like to take a stab at making it a bit more portable so that someone can download the project and have everything work.

For reference, I am using Ubuntu 16.04.2 LTS (I finally switched - life is simultaneously harder and easier).

**Comment by sebastian-c** [comment]
I made a branch called [boxrun](https://github.com/UNFAOstatistics/pocketbook_database/tree/boxrun) with a few changes that might make things run smoother. Essentially I made the paths dependent on a variable declared at the beginning.

Oh, and you really weren't kidding about the memory and time requirements - But it seems to be running without problems for the parts that I've run. I might let it run overnight and see how it goes.

**Comment by sebastian-c** [comment]
I let it run overnight - it gets stuck trying to read from `~/faosync/syb_bulk_database/` which doesn't seem to be created. Is `~/faosync/syb_bulk_database/metadata/meta_faostat.RDS` meant to be `~/faosync/pocketbook_database/local_data/faostat/metadata/meta_faostat.RDS` or symlinked to it?

**Comment by muuankarski** [comment]
Did you manage to get it running through? That path fix is required, I had those two files also at `syb_bulk_database`, but that location was no more updated.

Did two test runs and the bulks downloaded today from both FAO and WB ran ok.

**Comment by sebastian-c** [comment]
I haven't yet managed to get it running until the end. I've managed to do the bulk downloads ok, but I get stuck at some of the processing steps. My current issue is that I can't find the `faostat/rds/food_security_data_e_all_data_(norm).RDS` file. I can't work out when it's meant to be created.

**Comment by muuankarski** [comment]
I started a new fresh system and clone repos from github and try troubleshoot. Will get back later today as processing advances..

**Comment by sebastian-c** [comment]
Thanks, in the meantime I'll push the branches I have so you can see if I've taken a wrong turn in getting it to work.

I created a branch `faosyb_path` for my less certain changes regarding paths.

**Comment by sebastian-c** [comment]
Actually, that's not with your most recent changes. I'll merge in your master branch and see how it goes

**Comment by muuankarski** [comment]
Re-organised the whole lot to work standalone. "Should" work out-of-the box if both projects are cloned in `~/faosync/pocketbooks`. Have a look!

https://github.com/UNFAOstatistics/pocketbook_database/blob/master/README.md

**Comment by sebastian-c** [comment]
I'll give that a try after work and let it run overnight. I've abandoned my existing branches.

**Comment by sebastian-c** [comment]
I followed the instructions cloning into `~/faosync/pocketbooks`. Unfortunately, I forgot to properly log the output - I'll do so at lunch. I've summarised my last output:

```
...
OBESITY_BOTH
...
...
prod_val_meat.BASE.0406
Error in FUN(left, right): non-numeric argument to binary operator
Calls: scaleUnit -> Ops.data.frame -> eval -> eval
Execution halted
```

EDIT:

Ok, managed to find the problem - [run.R (line 330)](https://github.com/UNFAOstatistics/pocketbook_database/blob/106a226b520ed3ed629e4f3431a335d4977a7285/run.R#L330).

```
> traceback()
4: eval(expr, envir, enclos)
3: eval(f)
2: Ops.data.frame(df[, names(multiplierVec[inIndex])], multiplierVec.mat) at scaleUnit.R#54
1: scaleUnit(initial.df, meta.lst[["UNIT_MULT"]])

> sessionInfo()
R version 3.3.3 (2017-03-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.2 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=it_IT.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=it_IT.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=it_IT.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=it_IT.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] purrr_0.2.2      tidyr_0.4.1      tibble_1.2       ggplot2_2.2.0    tidyverse_1.1.1 
 [6] FAOSTAT_2.0      RJSONIO_1.3-0    readr_1.1.0      data.table_1.9.6 reshape2_1.4.1  
[11] plyr_1.8.1       dplyr_0.5.0     

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.4      forcats_0.2.0    class_7.3-11     tools_3.3.3      lubridate_1.6.0 
 [6] jsonlite_0.9.17  gtable_0.1.2     nlme_3.1-118     lattice_0.20-29  psych_1.7.3.21  
[11] rstudioapi_0.5   DBI_0.5-1        parallel_3.3.3   haven_1.0.0      e1071_1.6-8     
[16] xml2_1.0.0       stringr_1.0.0    httr_1.1.0       hms_0.3          classInt_0.1-23 
[21] grid_3.3.3       R6_2.1.2         readxl_0.1.1     foreign_0.8-61   modelr_0.1.0    
[26] magrittr_1.5     scales_0.4.1     MASS_7.3-35      rvest_0.3.2      assertthat_0.1  
[31] mnormt_1.5-5     colorspace_1.2-4 labeling_0.3     stringi_1.0-1    lazyeval_0.2.0  
[36] munsell_0.4.2    broom_0.4.2      chron_2.3-45    
```

**Comment by sebastian-c** [comment]
I found the source of the problem. It seems that the WDI data is being saved as character, meaning that the unit scaling fails. The problem is in `process_bulk`. Specifically [line 351](https://github.com/UNFAOstatistics/pocketbook_database/blob/106a226b520ed3ed629e4f3431a335d4977a7285/code/process_bulk.R#L351). Removing the `x62` column fixes the issue.

**Comment by sebastian-c** [comment]
Ok, after I made that fix, the whole thing ran from beginning to end. Good work, Markus.

**Comment by muuankarski** [comment]
I fixed the source of the problem now, if you pull the changes it should work without issues. however, the whole processing is one massive issue. A poorly implemented solution to a problem that does not exists. But that is another story! Have fun with it, don't burn your processors. And thanks for requesting it. i think this made the maintenance a lot more easy and flexible

**Comment by sebastian-c** [comment]
I managed to create the pdf as well! I tried to skimp out on some of dependencies, but I needed to install all of texlive in the end.

I'm glad you got it running - development of the project may need you but at least you're not needed to run it now.

I noticed the processor strain - but I only ever noticed it on one of them. I didn't see any parallelisation despite the fact that the scripts call themselves 'parallel'. I'm probably missing something.

**Comment by muuankarski** [comment]
As for parallelisation this line in run.R
https://github.com/UNFAOstatistics/pocketbook_database/blob/master/run.R#L721 fires up this script: https://github.com/UNFAOstatistics/pocketbook_database/blob/master/code/aggregate_functions/parallel_aggregate.sh that runs the four most time consuming tasks in parallel.

I sincerely hope that we could change the priority in the process from *finishing the books" into "finishing the software*. Then we could have a concise, purpose-build, tested and documented software for producing the books that would not require months of trial and error to learn how to configure and maintain it. After all, we are creating some promotional materials here! 

I would absolutely not mind if someone else would maintain once I would have the feeling that it is sustainable and that I wont be back on my plate in six months! In its current stage it would be abandoned for sure, as it will be totally nonfunctional in couple of weeks anyway as the underlying data changes.

But have a look around if you have any time. The `pockeboot`-system is re-written by me, and is slightly more purpose-built. Btw, you need to run the pdfs twice before the intermediate datas are populated and the books come out with proper content. Best of luck! ;)
